### ISSUE TITLE
ruby_block[config_replicaset] fails after initial chef-client run (Replication+Sharding) 

### ISSUE THREAD

**Issue by dquiles** [edelight/chef-mongodb#234]
1st run successfully installs and initiates db.

2nd run fails : 

```
================================================================================
Error executing action `create` on resource 'ruby_block[config_replicaset]'
================================================================================


NoMethodError
-------------
undefined method `[]' for nil:NilClass


Cookbook Trace:
---------------
/var/chef/cache/cookbooks/mongodb/libraries/mongodb.rb:102:in `configure_replicaset'
/var/chef/cache/cookbooks/mongodb/definitions/mongodb.rb:209:in `block (3 levels) in from_file'


Resource Declaration:
---------------------
# In /var/chef/cache/cookbooks/mongodb/definitions/mongodb.rb

207:     ruby_block 'config_replicaset' do
208:       block do
209:         MongoDB.configure_replicaset(new_resource.replicaset, replicaset_name, rs_nodes) unless new_resource.replicaset.nil?
210:       end
211:       action :nothing
212:     end
213:



Compiled Resource:
------------------
# Declared in /var/chef/cache/cookbooks/mongodb/definitions/mongodb.rb:207:in `block in from_file'

ruby_block("config_replicaset") do
  params {:mongodb_type=>"shard", :action=>[:enable, :start], :logpath=>"/var/log/mongodb", :dbpath=>"/var/lib/mongodb", :configserver=>[], :replicaset=>node[clientname], :notifies=>[], :port=>27017, :enable_rest=>false, :smallfiles=>false, :name=>"mongod"}
  action [:nothing]
  retries 0
  retry_delay 2
  block_name "config_replicaset"
  cookbook_name "mongodb"
  recipe_name "shard"
  block #<Proc:0x00000002972258@/var/chef/cache/cookbooks/mongodb/definitions/mongodb.rb:208>
end



[2014-02-06T04:21:07+00:00] ERROR: Running exception handlers
[2014-02-06T04:21:07+00:00] ERROR: Exception handlers complete
[2014-02-06T04:21:07+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
Chef Client failed. 19 resources updated
[2014-02-06T04:21:07+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

**Comment by pmacdougall** [comment]
Running into this too.  The problem lies on line 102 of the libraries/mongo.db:

``` ruby
server, port = Regexp.last_match[1].nil? ? ['localhost', node['mongodb']['port']] : Regexp.last_match[1].split(':')
```

Regexp.last_match is nil so referencing the 2nd element is causing the error.

I'm not sure why a last_match is expected to be there.  I can't find any other use of Regexp.

Just in case I missed something I kept the possibility of last_match[1] being populated with something but change the condition to check for nil and a length of at least 2:

``` patch
--- a/libraries/mongodb.rb
+++ b/libraries/mongodb.rb
@@ -99,7 +99,7 @@ class Chef::ResourceDefinitionList::MongoDB
     if result.fetch('ok', nil) == 1
       # everything is fine, do nothing
     elsif result.fetch('errmsg', nil) =~ /(\S+) is already initiated/ || (result.fetch('errmsg', nil) == 'already initialized')
-      server, port = Regexp.last_match[1].nil? ? ['localhost', node['mongodb']['port']] : Regexp.last_match[1].split(':')
+      server, port = Regexp.last_match.nil? || Regexp.last_match.length < 2 ? ['localhost', node['mongodb']['port']] : Regexp.last_match[1].split(':')
       begin
         connection = Mongo::Connection.new(server, port, :op_timeout => 5, :slave_ok => true)
       rescue
```

**Comment by dave-leblanc** [comment]
I was encountering this same issue and applied this patch and was able to successfully configure my server.

**Comment by pmacdougall** [comment]
FYI - I submitted a pull request and it has been merged already.

The pull request was issue 232, because I didn't know how to attach the PR to this issue.

**Comment by jamesonjlee** [comment]
you just reference the issue in the PR via #234

**Comment by josephholsten** [comment]
@jamesonjlee sounds like this can be closed

**Comment by jamesonjlee** [comment]
yep was fixed in at #242
